### DISCUSSION TITLE
Understanding the average utilisation for a FaaS service - any ideas?

### DISCUSSION THREAD

**Discussion by mrchrisadams** [Green-Software-Foundation/discussions#78] [Ask Anything]
Hi folks.

When we speak about serverless computing , it's often framed as inherently greener than other modes of computing, because some of the scaling happens in an  automated fashion, allowing for more efficient use of existing resources.

A new paper was published by Facebook  / Meta research with some interesting numbers, and it's linked below:
https://www.semanticscholar.org/paper/XFaaS%3A-Hyperscale-and-Low-Cost-Serverless-Functions-Sahraei-Demetriou/4bdec65474e2a5ff91fbfd9a130eb669ab4ea18d

> This paper presents the FaaS platform called XFaaS in Meta's hyperscale private cloud. XFaaS currently processes trillions of function calls per day on more than 100,000 servers. We describe a set of optimizations that help XFaaS achieve a daily average CPU utilization of 66%. Based on our anecdotal knowledge, this level of utilization might be several times higher than that of typical FaaS platforms.

I understand regular ol' compute utilisation _outside public cloud_ to be assumed to be in the low single digit percentages, and _inside public cloud_ to maybe be 20-30%, based on various claims I've seen in public comms about why the cloud is greener.

The paper linked above is interesting for loads of other reasons, but if you know of any other papers specifically referring to the utilisation efficiency of serverless, I'd welcome other links. 

If you have any other reckons or questions about the paper, you're very welcome to add them below (especially if remember to answer the initial question posed in this post...).

**Comment by kate-goldenring** [comment]
@mrchrisadams thank you for sharing this paper. Here is another [great paper by the original authors of Firecracker](https://www.usenix.org/system/files/nsdi20-paper-agache.pdf ) (AWS's microVMs for Lambda) that does a great job of discussing the characteristics of serverless and giving benchmarks.

The paper you share discusses how lengthy cold start times are a major limitation of serverless. Serverless or FaaS platforms are largely limited by the underlying unit of compute, say Firecracker MicroVMs in AWS Lambda's scenarios. The WebAssembly ecosystem is rapidly evolving and provides a path for a greener serverless, since Wasm has sub-millisecond start times (compared to 125ms lambdas) and can truly scale to 0. I'm an engineer at [Fermyon](https://www.fermyon.com/), which is a platform for running serverless WebAssembly applications, and wanted to answer the question of how many apps we could host on each of our nodes. We were able to scale up to supporting thousands of applications on a single [EC2 M5 instance](https://aws.amazon.com/ec2/instance-types/m5/). I demo this and discuss serverless Wasm more broadly in a [CNCF webinar](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cncf-on-demand-webinar-a-greener-cost-effective-cloud-with-serverless-webassembly/) I gave last week. 

So, to answer your initial question, i think serverless does provide a way to do more with less hardware, but i think it requires an optimal unit like Wasm.

**Reply by mrchrisadams** [reply]
Thanks @kate-goldenring  - woud you mind pointing me to some docs or a paper outlining how how isolation is managed between apps with WASM based tools like Fermyon / Spin?

I'm aware of how Fly's fantastic blog posts about how they manage isolation with MicroVMs
https://fly.io/blog/sandboxing-and-workload-isolation/

...and the v8 isolate approach used by Cloudflare
https://blog.cloudflare.com/cloud-computing-without-containers/

I understand that v8 isolates can run using Web Assembly, and if you can compile languages to WASM you could run them without needing the isolation offered by a container or microVM - is Fermyon using something along the lines of the isolate approach, or is it doing something more exotic like using Unikernels, as mentioned  in this discussion here?

https://github.com/orgs/Green-Software-Foundation/discussions/81

**Reply by kate-goldenring** [reply]
@mrchrisadams most of the isolation and safety guarantees are ensured by the Wasm runtime that you choose to use. Here is a great talk by Nick Fitzgerald that walks through how `wasmtime` (the most widely used runtime and the one we use in Spin/Fermyon) maintains it's security posture: https://youtu.be/FFPoOR_5urw?si=u0-iccThWbicY47j

Here is a paper that verifies the inability of wasm to escape it's sandbox (if instantiated securely by a runtime): https://cseweb.ucsd.edu//~dstefan/pubs/johnson:2021:veriwasm.pdf

One about how in the future, isolation can be further tightened with hardware mechanisms: https://cseweb.ucsd.edu/~dstefan/pubs/narayan:2023:hfi.pdf

And a newer paper studying compiler correctness' impact on isolation: https://www.cs.cornell.edu/~avh/veri-isle-preprint.pdf
